### ISSUE TITLE
Cache BestsellerReports for backend

### ISSUE THREAD

**Issue by Manuel-Innovapps** [nopSolutions/nopCommerce#6272]
nopCommerce version: 4.50.x +

Steps to reproduce the problem: 
1. Was found while debugging for #6271 
2. Our bestseller reports are taking forever to generate on azure, we don't know why yet.
3. Because of our long report generation times we saw that those reports are not being cached.

We propose to also cache the bestseller reports in admin, as we are doing in admin

**Comment by AndreiMaz** [comment]
@Manuel-Innovapps Duplicated https://github.com/nopSolutions/nopCommerce/issues/5792. Could you please apply the fix (commit) from this issue and let me know if it helps?

**Comment by Manuel-Innovapps** [comment]
Hey @AndreiMaz 

We just tested that out, but that didn't change anything for us. If we trust the Azure SQL Long Running queries tab, we can see clearly identify that the following statement is being generated and executed against azure:

```sql
SELECT
	Count(*)
FROM
	(
		SELECT
			0 as [c1]
		FROM
			[OrderItem] [orderItem_1]
				INNER JOIN [Order] [o] ON [orderItem_1].[OrderId] = [o].[Id]
				INNER JOIN [Product] [p] ON [orderItem_1].[ProductId] = [p].[Id]
				INNER JOIN [Address] [oba] ON [o].[BillingAddressId] = [oba].[Id]
		WHERE
			[o].[Deleted] = 0 AND [p].[Deleted] = 0
		GROUP BY
			[orderItem_1].[ProductId]
	) [t1]
```

And this clearly correlates with the following line (also this was the line that threw an exception, before we implemented #6271):
https://github.com/nopSolutions/nopCommerce/blob/804a1a97732fc01574c88ea8b37b9f0fc48ce9fd/src/Libraries/Nop.Services/Orders/OrderReportService.cs#L717

Which is what is causing problems for us.


The following SQL Statements might give you some details into our database:
``` sql
select count(*) from Product -- 1030
select count(*) from [Order] -- 19329
select count(*) from OrderItem -- 70184
select count(*) from Address -- 43923
```

We are running our shop on Azure with an MSSQL Database and the following details (equals to S0)
![image](https://user-images.githubusercontent.com/54266094/170983345-808e384d-be97-441b-9ed7-0973d4a40092.png)


Increasing the database tier might be an option to avoid this error. But the SQL Statement from above runs 40s+ on our database. So to get into an acceptable time, we would need to scale to S3 just to get the reports running. (Everything else runs fine)

At the moment, in order to avoid performance problems after our Go-Live (4.1 to 4.5 and onPrem to Azure) we opted to add an (hardcoded) filter to all bestseller reports and select only the orders made within the last year. This helped us dramatically reducing the time to something acceptable.

If you have any other idea on how we could improve this, please let us know.

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. 
Caching will not solve this problem. As part of ticket #5792, we carried out the maximum processing of this report and increased its performance several times, but this work was based on the information that we had at that time. 

Limiting the report to a certain period is a good option if it suits you. you can also add an additional index to the orders table:

CREATE NONCLUSTERED INDEX IX_Order_Delete_Id ON [Order] (Deleted);

Also, if possible, we would like to study the behavior of the request on your database, Please PM me a link to a backup copy of your database.

**Comment by Manuel-Innovapps** [comment]
Hey @skoshelev 

thanks for coming back to us.
I reverted the changes mentioned in the ticket description and initiated a deployment to our dev-systems After deployment I wasn't able to reproduce any of the mentioned behavior. So I rolled back our dev-systems. The main difference between now and then is that we were previously using single S0/S1/S2 databases on Azure. But because of our many databases (~ 20) we have since then migrated to an shared SQL Pool (S1 ~100 shared DTU between all databases) And this seems to be working quite well and apparently solves our initial problem for all our databases.

As for the database, due to the GPDR we are not allowed to send you this database. The only thing I could think of at the moment, is if we would be able to anonymize the database contents. Do you have any scripts for such operations available for 4.5? I remember we have on flying around for 3.9 or something, but this one destroyed some of our databases, so I'm not to sure about executing it.

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. if you're willing to share a backup, I'll write a script that will delete or replace all sensitive data, I don't think it will take long. we are glad that the problem is solved for you, however we will be very grateful for the opportunity to study the behavior on your data. Please indicate which data is sensitive to you other than information about customers.

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. I have written a generic script for cleaning sensitive data. You may want to expand it, it does not matter to us. I'm only using your database to find possible performance improvements for the BestsellerReports report, and then delete all the data.

```
DELETE FROM [ActivityLog]
GO
UPDATE [Address] SET
      [FirstName] = Null
      ,[LastName] = Null
      ,[Email] = Null
      ,[Company] = Null
      ,[City] = Null
      ,[Address1] = Null
      ,[Address2] = Null
      ,[PhoneNumber] = Null
      ,[FaxNumber] = Null
      ,[CustomAttributes] = Null
      ,[County] = Null
GO
UPDATE [Customer] SET [Username] = NUll
      ,[Email] = NUll
      ,[LastIpAddress] = NUll
      ,[EmailToRevalidate] = NUll
GO
UPDATE [CustomerPassword] SET 
      [Password] = Null
      ,[PasswordSalt] = Null
GO
DELETE FROM [GenericAttribute] WHERE [KeyGroup] = 'Customer'
GO
UPDATE [GiftCard] SET
      [RecipientName] = Null
      ,[RecipientEmail] = Null
      ,[SenderName] = Null
      ,[SenderEmail] = Null
GO
TRUNCATE TABLE [Log]
GO
TRUNCATE TABLE [OrderNote]
GO
TRUNCATE TABLE [QueuedEmail]
GO
```

I will close this ticket as the problem is not reproduced at the moment.

**Comment by skoshelev** [comment]
Closed #6272
